SLAM/PoseOptimizationSLAM3D/pose_optimization_slam_3d.py

```diff
@@ -0,0 +1,535 @@
+        qw = q.qw;qx = q.qx; qy = q.qy;qz = q.qz
+        dRdqw = 2 * np.array(
+            [[ qw, -qz,  qy],
+             [ qz,  qw, -qx],
+             [-qy,  qx,  qw]]
+        )
+        dRdqx = 2 * np.array(
+            [[ qx,  qy,  qz],
+             [ qy, -qx, -qw],
+             [ qz,  qw, -qx]]
+        )
+        dRdqy = 2 * np.array(
+            [[-qy,  qx,  qw],
+             [ qx,  qy,  qz],
+             [-qw,  qz, -qy]]
+        )
+        dRdqz = 2 * np.array(
+            [[-qz, -qw,  qx],
+             [ qw, -qz, -qy],
+             [ qz,  qy,  qz]]
+        )
+        dqdu = self.dQuat_dRV(rv)
+        dux = dRdqw * dqdu[0,0] + dRdqx * dqdu[1, 0] + dRdqy * dqdu[2, 0] + dRdqz * dqdu[3, 0]
+        duy = dRdqw * dqdu[0,1] + dRdqx * dqdu[1, 1] + dRdqy * dqdu[2, 1] + dRdqz * dqdu[3, 1]
+        duz = dRdqw * dqdu[0,2] + dRdqx * dqdu[1, 2] + dRdqy * dqdu[2, 2] + dRdqz * dqdu[3, 2]
+        return dux, duy, duz
+
+    def dRV_dQuat(self, q):
+
+        if 1 - q.qw**2 < 1e-7:
+            ret = np.array(
+            [[ 0.0, 2.0, 0.0, 0.0],
+             [ 0.0, 0.0, 2.0, 0.0],
+             [ 0.0, 0.0, 0.0, 2.0]]
+            )
+            return ret
+        c = 1/(1 - q.qw**2)
+        d = np.arccos(q.qw)/(np.sqrt(1-q.qw**2))
+        ret = 2.0 * np.array(
+            [[ c*q.qx*(d*q.qw-1),   d, 0.0, 0.0],
+             [ c*q.qy*(d*q.qw-1), 0.0,   d, 0.0],
+             [ c*q.qz*(d*q.qw-1), 0.0, 0.0,   d]]
+            )
+        return ret
+
+    def QMat(self, q):
+
+        qw = q.qw; qx = q.qx; qy = q.qy; qz = q.qz
+        Q = np.array(
+            [[ qw, -qx, -qy, -qz],
+             [ qx,  qw, -qz,  qy],
+             [ qy,  qz,  qw, -qx],
+             [ qz, -qy,  qx,  qw]]
+            )
+        return Q
+
+    def QMatBar(self, q):
+
+        qw = q.qw; qx = q.qx; qy = q.qy; qz = q.qz
+        Q = np.array(
+            [[ qw, -qx, -qy, -qz],
+             [ qx,  qw,  qz, -qy],
+             [ qy, -qz,  qw,  qx],
+             [ qz,  qy, -qx,  qw]]
+            )
+        return Q
+
+    def calc_error(self, pa, pb, t):
+
+        e0 = self.error_func(pa, pb, t)
+        Ja = np.identity(6); Jb = np.identity(6) 
+
+        rva_inv = pa.rv().inverted()
+        rotPaInv = rva_inv.toRotationMatrix()
+
+        Ja[:3,:3] = -rotPaInv
+        Jb[:3,:3] = rotPaInv
+
+        dRux, dRuy, dRuz = self.dR_dRV(rva_inv)
+
+        cvec = np.array([[pb.x - pa.x],[pb.y - pa.y],[pb.z - pa.z]])
+
+        Ja[0:3,3:4] = -dRux @ cvec
+        Ja[0:3,4:5] = -dRuy @ cvec
+        Ja[0:3,5:6] = -dRuz @ cvec
+
+        # rotation part: qdiff = qc-1 * qa-1 * qb
+        qainv = rva_inv.toQuaternion()
+        qcinv = t.rv().inverted().toQuaternion()
+        qb = pb.rv().toQuaternion()
+        qinvca = qcinv.quat_mult(qainv, out = 'Quaternion')
+        qdiff  = qinvca.quat_mult(qb, out = 'Quaternion')
+
+        Ja[3:6,3:6] = -self.dRV_dQuat(qdiff) @ self.QMat(qcinv) @ self.QMatBar(qb) @ self.dQuat_dRV(rva_inv)
```

Comment:
In this line has an error in my environment. The error occurs:

> Traceback (most recent call last):
>   File "pose_optimization_slam_3d.py", line 539, in <module>
>     main()
>   File "pose_optimization_slam_3d.py", line 525, in main
>     final_nodes = optimizer.optimize_path(nodes, consts, max_iter, min_iter)
>   File "pose_optimization_slam_3d.py", line 62, in optimize_path
>     graph_nodes, consts)
>   File "pose_optimization_slam_3d.py", line 98, in optimize_path_one_step
>     pa, pb, con.t)
>   File "pose_optimization_slam_3d.py", line 309, in calc_error
>     Ja[3:6,3:6] = -self.dRV_dQuat(qdiff) @ self.QMat(qcinv) @ self.QMatBar(qb) @ self.dQuat_dRV(rva_inv)
> TypeError: Object arrays are not currently supported

Reply:
I think this is because, the result of self.dQuad_RV(rva_inv) includes 1 element numpy array, see:
![SS 2019-09-07 at 13 57 16](https://user-images.githubusercontent.com/3813847/64470044-6d811400-d177-11e9-9081-b561a3d30bdb.png)

Reply:
The problem came from the nested numpy array, may be we need to just convert 1 size ndarray to a float.

Reply:
Oh, my python version is same as yours, 3.6.8. But numpy is slightly old, v1.15.4.

Reply:
How about this？
"""    
def dRV_dQuat(self, q):
        qw = q.qw[0]
        qx = q.qx[0]
        qy = q.qy[0]
        qz = q.qz[0]

        if 1 - qw**2 < 1e-7:
            ret = np.array(
            [[ 0.0, 2.0, 0.0, 0.0],
             [ 0.0, 0.0, 2.0, 0.0],
             [ 0.0, 0.0, 0.0, 2.0]]
            )
            return ret
        
        c = 1/(1 - qw**2)
        d = np.arccos(qw)/(np.sqrt(1-qw**2))
        ret = 2.0 * np.array(
            [[ c*qx*(d*qw-1),   d, 0.0, 0.0],
             [ c*qy*(d*qw-1), 0.0,   d, 0.0],
             [ c*qz*(d*qw-1), 0.0, 0.0,   d]]
            )
        return ret
"""

Reply:
Oh, it works!!. Could you please update it? It can support more older numpys.

Reply:
I downgraded numpy as 1.15.4 and found out that this problem ran,

Reply:
I updated it.